### ISSUE TITLE
Binding element incorrectly reported as having `any` type in JS with JSDoc starting with 5.6.2

### ISSUE THREAD

**Issue by burtek** [microsoft/TypeScript#59936]
### 🔎 Search Terms

Binding element implicitly has an any type 7031

### 🕗 Version & Regression Information

- This changed between versions 5.5.4 and 5.6.2

### ⏯ Playground Link

https://www.typescriptlang.org/play/?strictNullChecks=true&target=9&suppressImplicitAnyIndexErrors=true&ts=5.6.2&filetype=js#code/PQKhCgAIUgBAHAhgJ0QW0gbwPICMBWApgMYAuAvpANrED2AdgGYCWA5gLpQwIrpYAKKUs0QAbADwAlErWQATcQHJ8AZwaLIAH0jK19Yhu26GAVkUAaSCtLJm9VlXYA+J5RoMWrAHSI5c5sIMYgBizKKEKpzQwOCEAB7wsqSQjACu+oH0kPDIhEi5AMIebAAUmFCQlb7+mSFhEQBcWBWVrZCqDJAAvNRRrZQ9mOTgA1jkAJTNbXT0auFeorSsJR304yNAA

### 💻 Code

```js
/**
 * @param {Object} [config]
 * @param {Partial<Record<'json' | 'jsonc' | 'json5', string[]>>} [config.additionalFiles]
 */
export function prepareConfig({
    additionalFiles: {
        json = []
    } = {}
} = {}) {
    console.log(json)
}
```

### 🙁 Actual behavior

`json` is errored as `Binding element 'json' implicitly has an 'any[]' type.(7031)` even though it's typed as `string[]`

![image](https://github.com/user-attachments/assets/29b07884-02c7-4e10-bf6f-ee9a2b9b1967)

### 🙂 Expected behavior

No error expected, as `json` is typed as `string[]`

### Additional information about the issue

This worked well in 5.5.4 but broke with upgrade to 5.6.2. It works correctly in .ts file, but breaks in .js file with JSDoc.

**Comment by Andarist** [comment]
I'll fix this as part of https://github.com/microsoft/TypeScript/pull/59924

**Comment by burtek** [comment]
@Andarist ok, thanks. I guess it's gonna go out as 5.6 patch bump?

**Comment by Andarist** [comment]
@burtek in the end I decided to create a separate PR for this one since I was able to fix this independently and the other PR raises some design questions. The fix is available here: https://github.com/microsoft/TypeScript/pull/59975

As to your question, this is a recent regression so I'd advocate for backporting a fix but that's up to the TS team.

**Comment by ahejlsberg** [comment]
This issue actually existed before 5.6, but only for array binding patterns (playground [link](https://www.typescriptlang.org/play/?ts=5.5.4&filetype=js#code/PQKhCgAIUgBAHAhgJ0QW0gbwNrYM4AuyAlgHYDm2AugPy1UC+k2AHlVCMOAGYCupAYwLEA9qUjcATAApcAKzxjIAXmZUqKtRtXUAlFkjIApgV7JxCpQ3BA)):

```js
/**
 * @param {[[string[]?]?]} [x]
 */
function f1([[json = []] = []] = []) { return json } // Implicit any[] error on json binding element
```

In 5.6 we aligned array and object literal padding and the issue then showed up in object binding patterns as well:

```js
/**
 * @param {{ a?: { json?: string[] }}} [config]
 */
function f2({ a: { json = [] } = {} } = {}) { return json } // Implicit any[] error on json binding element
```

The culprit here is logic we have in `widenTypeInferredFromInitializer` to issue no implicit any errors in JS files. We shouldn't be issuing these errors when we're obtaining the implied type for a binding pattern. I'll put up a PR to fix. Unfortunately the fix in #59975 isn't the right one.
